### PR TITLE
CPM-1070: Save product identifier values

### PR DESCRIPTION
**Pull request by Balamung**
**Description (for Contributor and Core Developer)**

- Create the table to store product identifier values
- Fill this table on product save
- Introduce the IdentifierValue interface

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductIdentifiersSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    public function fillProductIdentifiers(GenericEvent $ev): void
+    {
+        $product = $ev->getSubject();
+        if (!$product instanceof ProductInterface) {
+            return;
+        }
+
+        $identifiers = \array_map(
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductIdentifiersSubscriber.php:40]: Add validation table exists

File: tests/back/Pim/Enrichment/Specification/Bundle/EventSubscriber/Product/OnSave/ComputeProductIdentifiersSubscriberSpec.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        $connection->executeStatement(
+            Argument::any(),
+            ['uuid' => $product->getUuid()->getBytes(), 'identifiers' => []],
+            Argument::any()
+        )->shouldBeCalled();
+
+        $this->fillProductIdentifiers($event);
+    }
+}
```
Comment (Balamung) [inline tests/back/Pim/Enrichment/Specification/Bundle/EventSubscriber/Product/OnSave/ComputeProductIdentifiersSubscriberSpec.php:99]: line

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductIdentifiersSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            StorageEvents::POST_SAVE => 'fillProductIdentifiers',
+        ];
+    }
+
+    public function fillProductIdentifiers(GenericEvent $ev): void
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductIdentifiersSubscriber.php:33]: Put in the same transaction inside the productSaver

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Product/SqlUpdateIdentifierValuesQuery.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * @copyright 2023 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class SqlUpdateIdentifierValuesQuery implements UpdateIdentifierValuesQuery
+{
+    public function __construct(private readonly Connection $connection)
+    {
+    }
+
+
```
Comment (Balamung) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Product/SqlUpdateIdentifierValuesQuery.php:23]: Double

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Db/CreateIdentifiersTableSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        $this->connection->executeStatement(
+            <<<SQL
+            CREATE TABLE IF NOT EXISTS pim_catalog_product_identifiers(
+                product_uuid BINARY(16) NOT NULL PRIMARY KEY,
+                identifiers JSON NOT NULL DEFAULT (JSON_ARRAY()),
+                CONSTRAINT pim_catalog_product_identifiers_pim_catalog_product_uuid_fk
+                    FOREIGN KEY (product_uuid) REFERENCES pim_catalog_product (uuid)
+                        ON DELETE CASCADE,
+                INDEX idx_identifiers ( (CAST(identifiers AS CHAR(255) ARRAY)) )
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Db/CreateIdentifiersTableSubscriber.php:38]: char (356) ?

**Review by cemorin — APPROVED** [review]

**Review by Beninho — APPROVED** [review]

**Review by antoinearrive — APPROVED** [review]

**Pull request merged**